### PR TITLE
Azure Communication Service - Phone Number Administration SDK

### PR DESCRIPTION
**Pull request by waynemo**
This PR adds the initial Phone Number Administration SDK. It includes the following:

- AutoRest Generated Client Code
- Convenience Layer Sync + Async Clients and Builder
- Unit Tests for Convenience Layer Builder
- Playback Tests for Convenience Layer Sync + Async Clients

To be completed in separate PRs:
- Exception handling / Null Checks
- Java Doc Code Snippets
- Long Running Operations
- README updates

### CODE REVIEW

**Review by JianpingChen — APPROVED** [review]

File: sdk/communication/azure-communication-administration/src/main/java/com/azure/communication/administration/implementation/PhoneNumberAdminClientImpl.java [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+
+    /**
+     * Initializes an instance of PhoneNumberAdminClient client.
+     *
+     * @param httpPipeline The HTTP pipeline to send requests through.
+     */
+    PhoneNumberAdminClientImpl(HttpPipeline httpPipeline, String endpoint) {
+        this.httpPipeline = httpPipeline;
+        this.endpoint = endpoint;
+        this.apiVersion = "2020-07-20-preview1";
```
Comment (ankitarorabit) [inline sdk/communication/azure-communication-administration/src/main/java/com/azure/communication/administration/implementation/PhoneNumberAdminClientImpl.java:80]: Shouldn't these get picked from some configuration rather than being hardcoded?

File: sdk/communication/azure-communication-administration/src/main/java/com/azure/communication/administration/implementation/PhoneNumberAdminClientImplBuilder.java [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+     * Builds an instance of PhoneNumberAdminClientImpl with the provided parameters.
+     *
+     * @return an instance of PhoneNumberAdminClientImpl.
+     */
+    public PhoneNumberAdminClientImpl buildClient() {
+        if (pipeline == null) {
+            this.pipeline =
+                    new HttpPipelineBuilder()
+                            .policies(new UserAgentPolicy(), new RetryPolicy(), new CookiePolicy())
+                            .build();
```
Comment (ankitarorabit) [inline sdk/communication/azure-communication-administration/src/main/java/com/azure/communication/administration/implementation/PhoneNumberAdminClientImplBuilder.java:59]: nit: I am not yet aware of the coding guidelines but the 'new' going to next line seems wrong

File: sdk/communication/azure-communication-administration/src/main/java/com/azure/communication/administration/implementation/PhoneNumberAdminClientImpl.java (hunk 1 above)
Comment (waynemo) [reply sdk/communication/azure-communication-administration/src/main/java/com/azure/communication/administration/implementation/PhoneNumberAdminClientImpl.java:80]: This code is generated by AutoRest.

**Review by ankitarorabit — APPROVED** [review]

File: sdk/communication/azure-communication-administration/src/main/java/com/azure/communication/administration/PhoneNumberAsyncClient.java [hunk 3]
```diff
@@ -0,0 +1,620 @@
...
+     * Gets a list of the supported area codes.
+     *
+     * @param locationType The type of location information required by the plan.
+     * @param countryCode The ISO 3166-2 country code.
+     * @param phonePlanId The plan id from which to search area codes.
+     * @param locationOptions A {@link List} of {@link LocationOptionsQuery} for querying the area codes.
+     * @return A {@link Mono} containing a {@link AreaCodes} representing area codes.
+     */
+    @ServiceMethod(returns = ReturnType.SINGLE)
+    public Mono<AreaCodes> getAllAreaCodes(
```
Comment (srnagar) [inline sdk/communication/azure-communication-administration/src/main/java/com/azure/communication/administration/PhoneNumberAsyncClient.java:82]: Please ensure that none of these methods throw exception. Exception should be returned through Mono.error or Flux.error.

File: sdk/communication/azure-communication-administration/src/main/java/com/azure/communication/administration/PhoneNumberAsyncClient.java [hunk 4]
```diff
@@ -0,0 +1,620 @@
...
+import com.azure.core.util.Context;
+import reactor.core.publisher.Mono;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+/**
+ * Asynchronous client for Communication service phone number operations
```
Comment (srnagar) [inline sdk/communication/azure-communication-administration/src/main/java/com/azure/communication/administration/PhoneNumberAsyncClient.java:44]: In general, our javadocs should also include codesnippets on how to instantiate the client at the class level and how to use the APIs for each of the public methods.

File: sdk/communication/azure-communication-administration/src/main/java/com/azure/communication/administration/PhoneNumberAsyncClient.java [hunk 5]
```diff
@@ -0,0 +1,620 @@
...
+     *
+     * @param phoneNumberCapabilitiesUpdate {@link Map} with the updates to perform
+     * @return A {@link Mono} containing
+     * a {@link UpdatePhoneNumberCapabilitiesResponse} representing the capabilities update.
+     */
+    @ServiceMethod(returns = ReturnType.SINGLE)
+    public Mono<UpdateNumberCapabilitiesResponse> updateCapabilities(
+        Map<PhoneNumber, NumberUpdateCapabilities> phoneNumberCapabilitiesUpdate) {
+        Map<String, NumberUpdateCapabilities> capabilitiesMap = new HashMap<>();
+        for (Map.Entry<PhoneNumber, NumberUpdateCapabilities> entry : phoneNumberCapabilitiesUpdate.entrySet()) {
```
Comment (srnagar) [inline sdk/communication/azure-communication-administration/src/main/java/com/azure/communication/administration/PhoneNumberAsyncClient.java:161]: Add null checks before using user-input (`phoneNumberCapabilitiesUpdate`) and provide helpful error message. Do the same for other methods too.

`Objects.requireNonNull(phoneNumberCapabilitiesUpdate, "'phoneNumberCapabilitiesUpdate' cannot be null");`

File: sdk/communication/azure-communication-administration/src/main/java/com/azure/communication/administration/PhoneNumberAsyncClient.java [hunk 6]
```diff
@@ -0,0 +1,620 @@
...
+/**
+ * Asynchronous client for Communication service phone number operations
+ */
+@ServiceClient(builder = PhoneNumberClientBuilder.class, isAsync = true)
+public final class PhoneNumberAsyncClient {
+
+    private final PhoneNumberAdminClientImpl phoneNumberAdminClient;
+
+    PhoneNumberAsyncClient(PhoneNumberAdminClientImpl phoneNumberAdminClient) {
+        this.phoneNumberAdminClient = phoneNumberAdminClient;
```
Comment (srnagar) [inline sdk/communication/azure-communication-administration/src/main/java/com/azure/communication/administration/PhoneNumberAsyncClient.java:52]: Looks like all APIs in this class only uses `phoneNumberAdminClient.getAdministrations()`. So, it might be simpler to have the `Administrations` type as instance variable.

File: sdk/communication/azure-communication-administration/src/main/java/com/azure/communication/administration/PhoneNumberClientBuilder.java [hunk 7]
```diff
@@ -0,0 +1,264 @@
...
+     * Sets the HTTP pipeline to use for the service client
+     * <p>
+     * If {@code pipeline} is set, all other settings aside from
+     * {@link PhoneNumberClientBuilder#endpoint(String) endpoint} are ignored.
+     *
+     * @param pipeline HttpPipeline to use
+     * @return The updated {@link PhoneNumberClientBuilder} object.
+     */
+    public PhoneNumberClientBuilder pipeline(HttpPipeline pipeline) {
+        this.pipeline = Objects.requireNonNull(pipeline, "'pipeline' cannot be null.");
```
Comment (srnagar) [inline sdk/communication/azure-communication-administration/src/main/java/com/azure/communication/administration/PhoneNumberClientBuilder.java:72]: This can be null in scenarios where the user initially set the pipeline but then decided to clear it out. These validations can be postponed until the build method is called.

File: sdk/communication/azure-communication-administration/src/main/java/com/azure/communication/administration/PhoneNumberClientBuilder.java [hunk 8]
```diff
@@ -0,0 +1,264 @@
...
+
+    /**
+     * Adds a policy to the set of existing policies that are executed after required policies.
+     *
+     * @param policy The retry policy for service requests.
+     * @return The updated {@link PhoneNumberClientBuilder} object.
+     * @throws NullPointerException If {@code policy} is {@code null}.
+     */
+    public PhoneNumberClientBuilder addPolicy(HttpPipelinePolicy policy) {
+        this.additionalPolicies.add(Objects.requireNonNull(policy, "'policy' cannot be null."));
```
Comment (srnagar) [inline sdk/communication/azure-communication-administration/src/main/java/com/azure/communication/administration/PhoneNumberClientBuilder.java:132]: Null check here is valid since we don't support adding null policies.

File: sdk/communication/azure-communication-administration/src/test/java/com/azure/communication/administration/PhoneNumberAsyncClientPlaybackTest.java [hunk 9]
```diff
@@ -0,0 +1,361 @@
...
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertNotNull;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+public class PhoneNumberAsyncClientPlaybackTest extends PhoneNumberPlaybackTestBase {
```
Comment (srnagar) [inline sdk/communication/azure-communication-administration/src/test/java/com/azure/communication/administration/PhoneNumberAsyncClientPlaybackTest.java:39]: Remove `Playback` from classname as it can be run in playback, record and live modes.

File: sdk/communication/azure-communication-administration/src/test/java/com/azure/communication/administration/PhoneNumberAsyncClientPlaybackTest.java [hunk 10]
```diff
@@ -0,0 +1,361 @@
...
+        assertEquals("ApplicationId", response.getValue().getPstnConfiguration().getApplicationId());
+    }
+
+    @Test()
+    public void unconfigureNumber() {
+        PhoneNumber number = new PhoneNumber(PHONENUMBER_TO_UNCONFIGURE);
+
+        Mono<Void> mono = this.getClient().unconfigureNumber(number);
+
+        mono.block();
```
Comment (srnagar) [inline sdk/communication/azure-communication-administration/src/test/java/com/azure/communication/administration/PhoneNumberAsyncClientPlaybackTest.java:322]: Do not block when testing async clients. Use `StepVerifier` instead.

File: sdk/communication/azure-communication-administration/src/test/java/com/azure/communication/administration/PhoneNumberClientBuilderTest.java [hunk 11]
```diff
@@ -0,0 +1,499 @@
...
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.atomic.AtomicReference;
+
+
+import static org.junit.jupiter.api.Assertions.*;
+import static org.mockito.Mockito.*;
+
+@Execution(value = ExecutionMode.SAME_THREAD)
```
Comment (srnagar) [inline sdk/communication/azure-communication-administration/src/test/java/com/azure/communication/administration/PhoneNumberClientBuilderTest.java:31]: Why is this required?

File: sdk/communication/azure-communication-administration/src/test/java/com/azure/communication/administration/PhoneNumberAsyncClientPlaybackTest.java (hunk 9 above)
Comment (waynemo) [reply sdk/communication/azure-communication-administration/src/test/java/com/azure/communication/administration/PhoneNumberAsyncClientPlaybackTest.java:39]: "PhoneNumberAsyncClientTest" is more of a unit test naming convention. Perhaps "PhoneNumberAsyncClientIntegrationTest"?

File: sdk/communication/azure-communication-administration/src/test/java/com/azure/communication/administration/PhoneNumberAsyncClientPlaybackTest.java (hunk 10 above)
Comment (waynemo) [reply sdk/communication/azure-communication-administration/src/test/java/com/azure/communication/administration/PhoneNumberAsyncClientPlaybackTest.java:322]: Will take a look at StepVerifier.

File: sdk/communication/azure-communication-administration/src/test/java/com/azure/communication/administration/PhoneNumberClientBuilderTest.java (hunk 11 above)
Comment (waynemo) [reply sdk/communication/azure-communication-administration/src/test/java/com/azure/communication/administration/PhoneNumberClientBuilderTest.java:31]: It's needed for mocking in the unit tests. If the tests are run in parallel, the mocks will have conflicts.

File: sdk/communication/azure-communication-administration/src/main/java/com/azure/communication/administration/PhoneNumberAsyncClient.java (hunk 3 above)
Comment (waynemo) [reply sdk/communication/azure-communication-administration/src/main/java/com/azure/communication/administration/PhoneNumberAsyncClient.java:82]: Perhaps we can update the error handling in a separate PR.

File: sdk/communication/azure-communication-administration/src/main/java/com/azure/communication/administration/PhoneNumberAsyncClient.java (hunk 5 above)
Comment (waynemo) [reply sdk/communication/azure-communication-administration/src/main/java/com/azure/communication/administration/PhoneNumberAsyncClient.java:161]: Perhaps we can update the error handling in a separate PR.

File: sdk/communication/azure-communication-administration/src/main/java/com/azure/communication/administration/PhoneNumberAsyncClient.java (hunk 6 above)
Comment (waynemo) [reply sdk/communication/azure-communication-administration/src/main/java/com/azure/communication/administration/PhoneNumberAsyncClient.java:52]: Good idea. Will update.

File: sdk/communication/azure-communication-administration/src/main/java/com/azure/communication/administration/PhoneNumberAsyncClient.java (hunk 4 above)
Comment (waynemo) [reply sdk/communication/azure-communication-administration/src/main/java/com/azure/communication/administration/PhoneNumberAsyncClient.java:44]: Let's add this in a separate PR.

File: sdk/communication/azure-communication-administration/src/main/java/com/azure/communication/administration/PhoneNumberClientBuilder.java (hunk 7 above)
Comment (waynemo) [reply sdk/communication/azure-communication-administration/src/main/java/com/azure/communication/administration/PhoneNumberClientBuilder.java:72]: To be addressed in separate PR.

File: sdk/communication/azure-communication-administration/src/test/java/com/azure/communication/administration/PhoneNumberAsyncClientPlaybackTest.java (hunk 10 above)
Comment (waynemo) [reply sdk/communication/azure-communication-administration/src/test/java/com/azure/communication/administration/PhoneNumberAsyncClientPlaybackTest.java]: Updated async client tests to use StepVerifier.
